### PR TITLE
ruler: add filtering support for prometheus rules api

### PR DESCRIPTION
**Pull request by ashwanthgoli**
**What this PR does / why we need it**:
Adds `rule_name`, `rule_group`, `file` and `type` query parameters for filtering the response of `/prometheus/api/v1/rules` endpoint. Replicates mimir's functionality: https://github.com/grafana/mimir/pull/5291

- all of them are optional.
- `type` paremeter accepts either `alert` or `record`
- `rule_name`, `rule_group`, `file` can accept multiple values and they filter the response accordingly.

There is a minor change in behavior: `/prometheus/api/v1/rules` endpoint will no longer return empty rule groups which is inline with both [prometheus](https://github.com/prometheus/prometheus/pull/12270/files#diff-315f251cdd7e93fcec1e7e9505744da1d1828f30d2b61d1f4ce963fa26bf1909R1403) and [mimir](https://github.com/grafana/mimir/pull/5291/files#diff-e5424c21c0e827bd1c9d3f669ed605897696bdc27993bc8bfd7113eba787b49dR1120). This is not a breaking change since rule groups with [no rules fail validation](https://github.com/grafana/loki/blob/27fbd62505f4412e3cb9180b1a5a66518bba9752/pkg/ruler/base/manager.go#L295) and cannot be created.

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/loki/issues/9295

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/ruler/base/api.go [hunk 1]
```diff
@@ -143,15 +152,34 @@ func (a *API) PrometheusRules(w http.ResponseWriter, req *http.Request) {
 	userID, err := tenant.TenantID(req.Context())
 	if err != nil || userID == "" {
 		level.Error(logger).Log("msg", "error extracting org id from context", "err", err)
-		respondError(logger, w, "no valid org id found")
+		respondServerError(logger, w, "no valid org id found")
 		return
 	}
 
-	w.Header().Set("Content-Type", "application/json")
```
Comment (kavirajk) [inline pkg/ruler/base/api.go:150]: Rationale for removing content type? Intentional?

Comment (ashwanthgoli) [reply pkg/ruler/base/api.go:150]: this was a duplicate, we set it later in the same function before returning

File: pkg/ruler/base/ruler.go [hunk 2]
```diff
@@ -799,26 +799,69 @@ func RemoveRuleTokenFromGroupName(name string) string {
...
+	if len(fs) == 0 {
+		return false
+	}
+	_, ok := fs[val]
+	return !ok
+}
+
+func (r *Ruler) getLocalRules(userID string, req *RulesRequest) ([]*GroupStateDesc, error) {
+	var (
+		getRecordingRules = true
```
Comment (kavirajk) [inline pkg/ruler/base/ruler.go:837]: I would prefer to create and switch these flags more direct way.

1. start with both false.
2. case AlertingRule => `alertRule=true`
3. case RecordingRule => `recordingRule=true`
4. case AnyRule => `alerRule=true; recordingRule=true`

This way it's more intuitive IMHO.

**Review by kavirajk — COMMENTED** [review]
LGTM. just one minor recommendation. 👍

File: pkg/ruler/base/ruler.go (hunk 2 above)
Comment (kavirajk) [reply pkg/ruler/base/ruler.go:837]: Also one other question. Do you think we also should support `exclude_alerts=<bool>`, just like prometheus rules endpoint does.
https://prometheus.io/docs/prometheus/latest/querying/api/#rules

Comment (ashwanthgoli) [reply pkg/ruler/base/ruler.go]: i'll add `exclude_alerts` in a follow-up

**Pull request merged**